File: benchmarks/bench.py
"""Benchmark for imager_client Python implementation.

Запуск из корня проекта:
    python docs/bench.py [iterations]

Измеряет клиентские операции (формирование URL/AssetType/HTML) — чистые
функции без HTTP. Админ-методы не бенчмаркастся.
"""
import os
import sys
import time

# Windows-консоли (cp1251 и др.) нужен UTF-8 для вывода "μs/op".
if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8")

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))

from src.imager import Imager  # noqa: E402

imager = Imager({"baseURL": "https://cdn.example.com/"})


def bench(name, fn, iterations=100000):
    start = time.perf_counter()
    for _ in range(iterations):
        fn()
    elapsed = time.perf_counter() - start
    ops = iterations / elapsed
    us_per_op = (elapsed / iterations) * 1_000_000
    print(f"{name}: {ops:,.0f} ops/s, {us_per_op:.2f} μs/op")


if __name__ == "__main__":
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 100_000

    bench("GetAsset", lambda: imager.GetAsset("image.jpg", "200x200", "webp"), n)
    bench(
        "GetAssets",
        lambda: imager.GetAssets("image.jpg", ["200x200", "400x400"], ["webp", "avif"]),
        n,
    )
    bench("GetAssetPath", lambda: imager.GetAssetPath("image.jpg", "200x200", "webp"), n)
    bench(
        "GetAssetsHtml",
        lambda: imager.GetAssetsHtml("image.jpg", ["200x200", "400x400"], ["webp", "avif"]),
        n,
    )

File: benchmarks/bench.php
<?php

/**
 * Benchmark for pkg-ru/imager-client PHP implementation.
 *
 * Запуск из корня проекта:
 *     php docs/bench.php [iterations]
 *
 * Измеряет клиентские операции (формирование URL/AssetType/HTML) — чистые
 * функции без HTTP. Админ-методы не бенчмаркастся.
 */

require_once __DIR__ . '/../vendor/autoload.php';

use imagerClient\Imager;

$imager = new Imager(['baseURL' => 'https://cdn.example.com/']);

function bench(string $name, callable $fn, int $iterations = 100000): void
{
    $start = hrtime(true);
    for ($i = 0; $i < $iterations; $i++) {
        $fn();
    }
    $elapsed = (hrtime(true) - $start) / 1e9;
    $ops = $iterations / $elapsed;
    $us = ($elapsed / $iterations) * 1_000_000;
    printf("%s: %s ops/s, %.2f μs/op\n", $name, number_format($ops, 0), $us);
}

$n = isset($argv[1]) ? (int) $argv[1] : 100_000;

bench("GetAsset", fn () => $imager->GetAsset("image.jpg", "200x200", "webp"), $n);
bench(
    "GetAssets",
    fn () => $imager->GetAssets("image.jpg", ["200x200", "400x400"], ["webp", "avif"]),
    $n
);
bench("GetAssetPath", fn () => $imager->GetAssetPath("image.jpg", "200x200", "webp"), $n);
bench(
    "GetAssetsHtml",
    fn () => $imager->GetAssetsHtml("image.jpg", ["200x200", "400x400"], ["webp", "avif"]),
    $n
);
